### PR TITLE
Add VerticalIntegralModel for integrating flow and tendency

### PR DESCRIPTION
**Pull request by blallen**
# Description

This PR adds a VerticalIntegralModel that can be used to integrate both the flow field and flow tendency for use in the split explicit model in #1268. This model is tested for correctness using the 2D and 3D hydrostatic spindown initial states.

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Arrays/MPIStateArrays.jl [hunk 1]
```diff
@@ -515,12 +515,27 @@ function LinearAlgebra.norm(
...
-        locnorm = weighted_norm_impl(Q.realdata, W, Val(p), dims)
+
+        # Check that the arrays are the right dimensionality
+        @assert ndims(realdata) in (2, 3)
+
+        # Reshape to 3 dimensions is necessary
+        if ndims(realdata) == 2
+            realdata =
+                reshape(realdata, size(realdata, 1), 1, size(realdata, 2))
+        end
```
Comment (jkozdon) [inline src/Arrays/MPIStateArrays.jl:530]: The other option here is to force the user to send in the data correctly shaped, but then they cannot use the `vars` pull-out from the `MPIStateArray`

File: test/Ocean/SplitExplicit/test_vertical_integral_model.jl [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            ∫u = reshape(flat_∫u, Nq^2, number_auxiliary, nhorzrealelem)
+
+            U = @view Q_2D[:, index_2D, :]
+
+            error =
+                euclidean_distance(Q_2D, Q_int; ArealQ = U, BrealQ = ∫u) /
+                norm(Q_2D; realdata = U)
+
+            println("error = ", error)
+            @test isapprox(error, FT(0.0); atol = 1e-12)
```
Comment (jkozdon) [inline test/Ocean/SplitExplicit/test_vertical_integral_model.jl:199]: I like that there are tests!

File: test/Ocean/SplitExplicit/test_vertical_integral_model.jl [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+function test_vertical_integral_model()
+    mpicomm = MPI.COMM_WORLD
+    ArrayType = ClimateMachine.array_type()
+
+    ll = uppercase(get(ENV, "JULIA_LOG_LEVEL", "INFO"))
+    loglevel = ll == "DEBUG" ? Logging.Debug :
+        ll == "WARN" ? Logging.Warn :
+        ll == "ERROR" ? Logging.Error : Logging.Info
+    logger_stream = MPI.Comm_rank(mpicomm) == 0 ? stderr : devnull
+    global_logger(ConsoleLogger(logger_stream, loglevel))
```
Comment (jkozdon) [inline test/Ocean/SplitExplicit/test_vertical_integral_model.jl:94]: I don't think this is needed anymore, it's done in [`Driver.jl`](https://github.com/CliMA/ClimateMachine.jl/blob/1d183062da2687e4e3255a0302c22e21df8c1020/src/Driver/Driver.jl#L458-L470)

```suggestion
```

Comment (blallen) [reply test/Ocean/SplitExplicit/test_vertical_integral_model.jl:94]: Does this apply even if we aren't using driver? Is it still handled in `Clima.init()`?

I don't want to add the split-explicit setup to the Driver framework until we know what we actually need to do

File: test/Ocean/SplitExplicit/test_vertical_integral_model.jl (hunk 2 above)
Comment (blallen) [reply test/Ocean/SplitExplicit/test_vertical_integral_model.jl:199]: Tests are great!

File: test/Ocean/SplitExplicit/test_vertical_integral_model.jl (hunk 3 above)
Comment (jkozdon) [reply test/Ocean/SplitExplicit/test_vertical_integral_model.jl:94]: > Does this apply even if we aren't using driver? Is it still handled in `Clima.init()`?

I believe so. @kpamnany?

Comment (kpamnany) [reply test/Ocean/SplitExplicit/test_vertical_integral_model.jl:94]: [Yes](https://github.com/CliMA/ClimateMachine.jl/blob/master/src/Driver/Driver.jl#L457-L470).

File: src/Arrays/MPIStateArrays.jl (hunk 1 above)
Comment (blallen) [reply src/Arrays/MPIStateArrays.jl:530]: @jkozdon looks like one of the tests for this failed, can you take a look?

https://github.com/CliMA/ClimateMachine.jl/runs/831119353#step:7:235

File: test/Ocean/SplitExplicit/test_vertical_integral_model.jl (hunk 3 above)
Comment (blallen) [reply test/Ocean/SplitExplicit/test_vertical_integral_model.jl:94]: okay great, I'll remove them

**Comment by blallen** [comment]
bors try

File: src/Arrays/MPIStateArrays.jl (hunk 1 above)
Comment (jkozdon) [reply src/Arrays/MPIStateArrays.jl:530]: > @jkozdon looks like one of the tests for this failed, can you take a look?

Forgot to test in a parallel. Pushed a fix.

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
@jkozdon how do you want to squash this? one commit for the VerticalIntegralModel stuff and one commit for the changes to MPIStateArrays (and new tests for those changes)?

**Comment by jkozdon** [comment]
@jkozdon<https://github.com/jkozdon> how do you want to squash this? one commit for the VerticalIntegralModel stuff and one commit for the changes to MPIStateArrays (and new tests for those changes)?

Squashing is fine. You want me to do it or will you?

**Comment by blallen** [comment]
> @jkozdon<https://github.com/jkozdon> how do you want to squash this? one commit for the VerticalIntegralModel stuff and one commit for the changes to MPIStateArrays (and new tests for those changes)? Squashing is fine. You want me to do it or will you?

I can do it, I just wanted to make sure you're okay with the separation (since the MPIStateArray stuff shouldn't be coupled with the VerticalIntegralModel stuff).

**Comment by blallen** [comment]
@jkozdon it looks like one of the tests wasn't GPU friendly

https://gist.github.com/climabot/e36ba5be2df904b6cae58584c0956575#file-out_10287711

**Comment by jkozdon** [comment]
> @jkozdon it looks like one of the tests wasn't GPU friendly
> 
> https://gist.github.com/climabot/e36ba5be2df904b6cae58584c0956575#file-out_10287711

Yeah, most of them actually weren't.

This shows the problem:
```julia
using CUDA
using LazyArrays
CUDA.allowscalar(false)
A = CuArray(rand(10, 11, 12));
B = @view A[:, 1, :];
@show mapreduce(abs, +, B, dims = :)
```
which produces:
```
julia> using CUDA

julia> CUDA.allowscalar(false)

julia> A = CuArray(rand(10, 11, 12));

julia> B = @view A[:, 1, :];

julia> @show mapreduce(abs, +, B, dims = :)
ERROR: scalar getindex is disallowed
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] assertscalar(::String) at /home/jekozdon/.julia/packages/GPUArrays/X4SqE/src/host/indexing.jl:41
 [3] getindex at /home/jekozdon/.julia/packages/GPUArrays/X4SqE/src/host/indexing.jl:96 [inlined]
 [4] _getindex at ./abstractarray.jl:1003 [inlined]
 [5] getindex at ./abstractarray.jl:980 [inlined]
 [6] getindex at ./subarray.jl:236 [inlined]
 [7] _getindex at ./abstractarray.jl:1020 [inlined]
 [8] getindex at ./abstractarray.jl:980 [inlined]
 [9] iterate at ./abstractarray.jl:913 [inlined]
 [10] iterate at ./abstractarray.jl:911 [inlined]
 [11] _foldl_impl(::Base.MappingRF{typeof(abs),Base.BottomRF{typeof(+)}}, ::Base._InitialValue, ::SubArray{Float64,2,CuArray{Float64,3,Nothing},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}}},false}) at ./reduce.jl:53
 [12] foldl_impl at ./reduce.jl:45 [inlined]
 [13] mapfoldl_impl(::typeof(abs), ::typeof(+), ::NamedTuple{(),Tuple{}}, ::SubArray{Float64,2,CuArray{Float64,3,Nothing},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}}},false}) at ./reduce.jl:41
 [14] mapfoldl(::Function, ::Function, ::SubArray{Float64,2,CuArray{Float64,3,Nothing},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}}},false}; kw::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./reduce.jl:157
 [15] mapfoldl at ./reduce.jl:157 [inlined]
 [16] _mapreduce at ./reduce.jl:413 [inlined]
 [17] _mapreduce_dim at ./reducedim.jl:312 [inlined]
 [18] #mapreduce#580 at ./reducedim.jl:307 [inlined]
 [19] top-level scope at show.jl:613
```

Not sure how to fix it, maybe @mwarusz or @vchuravy can help.

**Comment by jkozdon** [comment]
The euclidean distance can be fixed by requiring you always have 3-d arrays and disallowing the 2-D array case.

**Comment by jkozdon** [comment]
Also, `test/Ocean/SplitExplicit/test_vertical_integral_model.jl` fails on the GPU as well.

**Comment by jkozdon** [comment]
> The euclidean distance can be fixed by requiring you always have 3-d arrays and disallowing the 2-D array case.

Pushed partial fix. Feel free to drop it if you do not like it.

**Comment by blallen** [comment]
> > The euclidean distance can be fixed by requiring you always have 3-d arrays and disallowing the 2-D array case.
> 
> Pushed partial fix. Feel free to drop it if you do not like it.

If this works, I'm okay to go ahead with it. Going to give bors a try

**Comment by blallen** [comment]
bors try

**Comment by blallen** [comment]
> ## try
> Build failed:
> 
> * [ci/slurmci](https://gist.github.com/28404e70e2bb753a2e16d4da629c76ba)

@jkozdon did you expect these tests to fail?

**Comment by leios** [comment]
The issue here is that `mapreduce!()` (line 707 in `MPIStateArrays.jl`) is not defined for a SubArray of CuArray:

```
julia> f = abs2
abs2 (generic function with 6 methods)

julia> op = +
+ (generic function with 280 methods)

julia> ca = CuArray(ones(100,100));

julia> cc = @view ca[1:10, 1:10];

julia> mapreduce(f, op, cc, dims = :);
ERROR: scalar getindex is disallowed
Stacktrace:
 [1] assertscalar(::String) at /home/leios/.julia/packages/GPUArrays/X4SqE/src/host/indexing.jl:41
 [2] getindex at /home/leios/.julia/packages/GPUArrays/X4SqE/src/host/indexing.jl:96 [inlined]
 [3] _getindex at ./abstractarray.jl:1003 [inlined]
 [4] getindex at ./abstractarray.jl:980 [inlined]
 [5] getindex at ./subarray.jl:236 [inlined]
 [6] _getindex at ./abstractarray.jl:1020 [inlined]
 [7] getindex at ./abstractarray.jl:980 [inlined]
 [8] iterate at ./abstractarray.jl:913 [inlined]
 [9] iterate at ./abstractarray.jl:911 [inlined]
 [10] _foldl_impl(::Base.MappingRF{typeof(abs2),Base.BottomRF{typeof(+)}}, ::Base._InitialValue, ::SubArray{Float64,2,CuArray{Float64,2,Nothing},Tuple{UnitRange{Int64},UnitRange{Int64}},false}) at ./reduce.jl:53
 [11] foldl_impl at ./reduce.jl:45 [inlined]
 [12] mapfoldl_impl(::typeof(abs2), ::typeof(+), ::NamedTuple{(),Tuple{}}, ::SubArray{Float64,2,CuArray{Float64,2,Nothing},Tuple{UnitRange{Int64},UnitRange{Int64}},false}) at ./reduce.jl:41
 [13] mapfoldl(::Function, ::Function, ::SubArray{Float64,2,CuArray{Float64,2,Nothing},Tuple{UnitRange{Int64},UnitRange{Int64}},false}; kw::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at ./reduce.jl:157
 [14] mapfoldl at ./reduce.jl:157 [inlined]
 [15] _mapreduce at ./reduce.jl:413 [inlined]
 [16] _mapreduce_dim at ./reducedim.jl:312 [inlined]
 [17] #mapreduce#580 at ./reducedim.jl:307 [inlined]
 [18] top-level scope at REPL[26]:1
```

Hope it helps! Maybe Adapt?

**Comment by blallen** [comment]
To speed this up, I can Use a state check to test the integral model on CI and we can split off the MPI State Arrays portion into its own PR.

Does that sound reasonable, @jkozdon?

**Comment by jkozdon** [comment]
> Does that sound reasonable, @jkozdon?

Whatever you think is best.

As @leios and my examples above show, the problem is that the mapreduce doesn't work on the cuarray subarrays. Until someone sorts that out, this won't work.

**Comment by blallen** [comment]
Okay, split out the MPIStateArray stuff in #1344. This PR should pass all tests now.

**Comment by jkozdon** [comment]
> Build failed:
> 
> * [docs-build](https://github.com/CliMA/ClimateMachine.jl/runs/870746414)

The log shows that one of the links being used is dead
- https://github.com/CliMA/ClimateMachine.jl/runs/870746414#step:4:4
- http://azure.archive.ubuntu.com/ubuntu/pool/main/m/mesa/libegl-mesa0_19.2.8-0ubuntu0~18.04.3_amd64.deb

**Comment by blallen** [comment]
> > Build failed:
> > 
> > * [docs-build](https://github.com/CliMA/ClimateMachine.jl/runs/870746414)
> 
> The log shows that one of the links being used is dead
> 
> * https://github.com/CliMA/ClimateMachine.jl/runs/870746414#step:4:4
> * http://azure.archive.ubuntu.com/ubuntu/pool/main/m/mesa/libegl-mesa0_19.2.8-0ubuntu0~18.04.3_amd64.deb

@charleskawczynski could you take a look?

**Comment by jkozdon** [comment]
> The log shows that one of the links being used is dead

FYI: The link is being called by `apt-get`, not anything in ClimateMachine / Julia

**Pull request merged**